File: praca-inzynierska-frontend/src/app/app/model/dto/StatisticsPositionYearDto.ts
import { PositionDto } from './PositionDto';
import { StatisticsDto } from './StatisticsDto';
import { StatisticsEmployeeYearDto } from './StatisticsEmployeeYearDto';

export class StatisticsPositionYearDto {

    constructor(
        public position: PositionDto,
        public employees: StatisticsEmployeeYearDto[],
        public totalMonths: StatisticsDto[],
        public total: StatisticsDto
    ) {}

    static copyOf(other: StatisticsPositionYearDto): StatisticsPositionYearDto {
        return new StatisticsPositionYearDto(other.position, other.employees.map(e => StatisticsEmployeeYearDto.copyOf(e)),other.totalMonths.map(m => StatisticsDto.copyOf(m)), StatisticsDto.copyOf(other.total))
    }

}

File: praca-inzynierska-backend/application/src/main/kotlin/jwer/backend/model/dto/StatisticsPositionYearDto.kt
package jwer.backend.model.dto

import jwer.backend.model.entity.Shift

data class StatisticsPositionYearDto(
        val position: PositionDto,
        val employees: List<StatisticsEmployeeYearDto> = listOf(),
        val totalMonths: List<StatisticsDto> = Array(12) {StatisticsDto(it)}.toList(),
        val total: StatisticsDto = StatisticsDto(null)
) {

    fun add(shift: Shift) {
        val month = shift.period.start.monthValue - 1
        totalMonths[month] += shift
        total += shift
        employees.find { it.employee.employeeId == shift.employee.id }?.add(shift)
    }

}